tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,367 @@
+        coll_recomb_coeff,
+        partition_function,
+        levels,
+        level_boltzmann_factor,
+        phi,
+        rate_matrix_index,
+        number_density,
+    ):
+        """Calculates ion number densities and electron densities using NLTE ionization.
+
+        Parameters
+        ----------
+        gamma : DataFrame
+        alpha_sp : DataFrame
+        alpha_stim : DataFrame
+        coll_ion_coeff : DataFrame
+        coll_recomb_coeff : DataFrame
+        partition_function : DataFrame
+        levels : MultiIndex
+        level_boltzmann_factor : DataFrame
+        phi : DataFrame
+        rate_matrix_index : MultiIndex
+        number_density : DataFrame
+
+        Returns
+        -------
+        ion_number_densities_nlte: DataFrame
+        electron_densities_nlte: Series
+        """
+
+        (
+            total_photo_ion_coefficients,
+            total_rad_recomb_coefficients,
+            total_coll_ion_coefficients,
+            total_coll_recomb_coefficients,
+        ) = self.prepare_ion_recomb_coefficients_nlte_ion(
+            gamma,
+            alpha_sp,
+            alpha_stim,
+            coll_ion_coeff,
+            coll_recomb_coeff,
+            partition_function,
+            levels,
+            level_boltzmann_factor,
+        )
+
+        # >>>TODO:initial electron density should be included in the initial guess, added in a future PR
+        initial_electron_density = number_density.sum(axis=0)
+        # <<<
+        rate_matrix = self.calculate_rate_matrix(
+            phi[0],
+            initial_electron_density[0],
+            rate_matrix_index,
+            total_photo_ion_coefficients[0],
+            total_rad_recomb_coefficients[0],
+            total_coll_ion_coefficients[0],
+            total_coll_recomb_coefficients[0],
+        )
+        return (
+            -1,
+            -1,
+        )  # function still empty, that's why return statement is arbitrary at this point
+
+    @staticmethod
+    def calculate_rate_matrix(
+        phi_shell,
+        electron_density,
+        rate_matrix_index,
+        total_photo_ion_coefficients,
+        total_rad_recomb_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """
+
+        Parameters
+        ----------
+        phi_shell : DataFrame
+            Saha Factors in the current shell
+        electron_density : float
+            Guess for electron density in the current shell
+        rate_matrix_index : MultiIndex
+            Index used for constructing the rate matrix
+        total_photo_ion_coefficients : DataFrame
+            Photo ionization coefficients
+        total_rad_recomb_coefficients : DataFrame
+            Radiative recombination coefficients(should get multiplied by electron density)
+        total_coll_ion_coefficients : DataFrame
+            Collisionional ionization coefficients(should get multiplied by electron density)
```

Comment:
extra ion sneaked its way into the description on this line :)

Reply:
Found it, got rid of it :)